ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+handle = client.start_workflow(SumWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+handle.result # => 3
+```
+
+#### Alternative approaches
+
+- Starting a foreign workflow from a string name
+
+```ruby
+client.start_workflow('sum-workflow', 1, 2, id: 'id-1', task_queue: 'demo')
+```
+
+- Starting a foreign workflow with the help of a stub
+
+```ruby
+class MyWorkflow < Temporal::Workflow
+  # Optionally override the name of the workflow
+  workflow_name 'my.workflow'
+end
+
+client.start_workflow(MyWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
+```
+
+
+## Start a worker
+
+```ruby
+require 'temporal/worker'
+require 'sum_workflow'
+require 'some_other_workflow'
+require 'sum_activity'
+require 'some_other_activity'
+
+connection = Temporal::Connection.connect('localhost:7233')
+worker = Temporal::Worker.new(
+  connection,
+  namespace: 'demo',
+  task_queue: 'demo',
+  workflows: [SumWorkflow, SomeOtherWorkflow],
+  activities: [SumActivity, SomeOtherActivity]
+)
+
+worker.run
+```
+
+#### Alternative approach
+
+To support passing additional context to the activity classes a worker can be initialised with an
+instance of the activity class (only available for activities).
+
+```ruby
+redis = Redis.new('localhost:6379')
+
+Temporal::Worker.new(
+  ...,
+  activities: [CalculatorActivities.new(redis)]
+)
+```
+
+*NOTE: This however will come with a heavy warning — the activity must be strictly thread-safe as
+the provided instance will be called from multiple threads (and possibly at the same exact time).*
+
+
+## Execute a child workflow
+
+`parent_workflow.rb`
+
+```ruby
+class ChildWorkflow < Temporal::Workflow
+  def execute(name)
+    "Hello, #{name}!"
+  end
+end
+
+class ParentWorkflow < Temporal::Workflow
+  def execute
+    workflow.execute_child_workflow(ChildWorkflow, 'parent', id: 'id-2')
+  end
+end
+```
+
+
+## Async workflow
+
+Ruby doesn't have a standard (or at least very popular) framework for async execution of the code.
+Because of this we can assume that most Ruby engineers using the SDK won't be immediately familiar
+with writing async code. Because of this we have decided to make async execution opt-in, keeping
+everything synchronous by default.
+
+We have also chosen an eager execution pattern where any `async` block is started immediately and
+performed until blocked before proceeding with the execution. This is a bit easier to get a hold of
+and eliminates the risk of forgetting to actually start an async block. However this means that
+we'll need to find a solution for unhandled promise rejections, which we'll cover separately in a
+more detailed workflow proposal.
+
+We will introduce a helper method `async` to execute a block of code asynchronously:
+
+```ruby
+require 'sum_activity'
+
+class AsyncWorkflow < Temporal::Workflow
+  def execute
+    activity_promise = async { workflow.execute_activity(SumActivity, 1, 2) }
+    workflow_promise = async { workflow.execute_child_workflow(ChildWorkflow, 'parent') }
```

Comment:
Yes, inside the block itself is synchronous (unless you nest `async`). So `execute_child_workflow` twice in the same async block just means that the whole block is async/background, but the second won't execute until after the first is complete.

Reply:
Yep, what Chad said ^. Basically:


```async
async do
  workflow.execute_activity(Activity1)
  
  promise_1 = async do
    workflow.execute_activity(Activity2)
    workflow.execute_activity(Activity3)
  end

  promise_2 = async { workflow.execute_activity(Activity4) }
  
  async.all(promise_1, promise_2).await
end
```

This will not start `Activity2` and `Activity4` until `Activity1` has completed. Then start `Activity2` and `Activity4` at the same time, but not `Activity3`. After `Activity2` is completed, `Activity3` will get executed. And the whole outer `async` block will complete after all activities have completed. Hope that makes sense